shell: uart: Add option to route shell through USB CDC ACM

KubaFYI opened the pull request:
Addition of a simple way of permitting routing of shell through a USB CDC ACM UART.

Fixes: #26051.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

KubaFYI commented:
When wanting to route the shell through USB CDC ACM there is a problem with the shell getting possibly initialised before the CDC ACM driver. 

I managed to successfully accomplish what I need by adding a call to `usb_enable()` in the `enable_shell_uart()` function and changing the literal `0` shell initialisation priority to `CONFIG_APPLICATION_INIT_PRIORITY` (so that it happens after the USB device's `CONFIG_KERNEL_INIT_PRIORITY_DEVICE`.

However, I'm not sure if this is the correct way to do it with me being possibly unaware of the reason why shell's initialisation needs to be `0`, an the best place to `usb_enable()` - I'm open to suggestions of better ways of doing it.

KubaFYI commented:
Also, the way I defined Kconfig for this option doesn't correctly source the CDC ACM UART device name from its Kconfig value. I'm pretty new to Kconfig so would appreciate any insights on this one.

jakub-uC commented on subsys/shell/Kconfig.backends:
```diff
@@ -30,8 +30,19 @@ config SHELL_PROMPT_UART
 
+config UART_SHELL_USE_CDC_ACM
+	bool "Deliver Shell through USB CDC AMC"
+	default n
```
This line is obsolete, `n` is a default choice.

jfischer-phytec-iot replied:
I think changes in Kconfig.backends are not necessary at all. Adding CONFIG_SHELL_ON_DEV_NAME="CDC_ACM_0" to the configuration overlay/file of an application/sample should also work.

KubaFYI replied:
The only reason I introduced the new key in Kconfig was because a change to `SYS_INIT` in `shell_uart.c` has to be done at compile time (and then the same key can also control whether `usb_enable` is triggered or not. Do you have any idea on how this can be accomplished differently?

jakub-uC commented on samples/subsys/shell/shell_module/prj.conf:
```diff
@@ -10,3 +10,7 @@ CONFIG_THREAD_NAME=y
 CONFIG_DEVICE_SHELL=y
 CONFIG_POSIX_CLOCK=y
 CONFIG_DATE_SHELL=y
+CONFIG_USB_DEVICE_STACK=y
```
Do we want to enable it by default in this example?

nordic-krch replied:
i think it should only be enabled in board specific conf file.

jfischer-phytec-iot replied:
or configuration overlay

bpbradley commented:
Just a comment, but I was in the process of implementing something similar and saw this PR. I tried yours and it works perfectly 👍  

Hopefully you get some feedback on the finer details and it it is merged soon. The only thing I might suggest is to probably move the sample configs into a separate conf file. Maybe similar to how samples/usb/cdc_acm does it with overlays. I also needed to add CONFIG_USB=y to the sample.

nordic-krch commented on subsys/shell/shell_uart.c:
```diff
@@ -273,7 +275,9 @@ static int enable_shell_uart(struct device *arg)
 		(CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL > LOG_LEVEL_DBG) ?
 		CONFIG_LOG_MAX_LEVEL : CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL;
-
+#if CONFIG_UART_SHELL_USE_CDC_ACM
```
use `if (IS_ENABLED(CONFIG_UART_SHELL_USE_CDC_ACM) {`

nordic-krch commented on subsys/shell/shell_uart.c:
```diff
@@ -282,7 +286,11 @@ static int enable_shell_uart(struct device *arg)
+#if CONFIG_UART_SHELL_USE_CDC_ACM
+SYS_INIT(enable_shell_uart, POST_KERNEL, CONFIG_APPLICATION_INIT_PRIORITY);
+#else
 SYS_INIT(enable_shell_uart, POST_KERNEL, 0);
```
just do:
```
SYS_INIT(enable_shell_uart, POST_KERNEL, IS_ENABLED(CONFIG_UART_SHELL_USE_CDC_ACM) ? CONFIG_APPLICATION_INIT_PRIORITY : 0);
```

emob-nordic commented:
Hi. I think this PR is worth pursuing. I have some concerns.
1) Should we edit ```shell_uart.c``` or create another for cdc_acm like: ```shell_cdc_acm.c```? @nordic-krch @jakub-uC.
2) I think Kconfig.backend should have separate config for CDC_ACM configuration. For now it seems to be shared across with UART. For instance something obvious as prompt could be ``` cdc_acm:~$ ```.
3) LOGs are directed to the cdc_acm shell as well. Is it the behaviour You wanted?

I will have a deeper look into this PR, this is just first impression.

bpbradley commented:
I am wondering if this PR has any plans to move forward. I have my own solution working for now, but am hoping this can be upstreamed.

KubaFYI commented:
@bpbradley I was quite busy lately but I am planning to address the concerns here and move the PR forward.

bpbradley commented:
> @bpbradley I was quite busy lately but I am planning to address the concerns here and move the PR forward.

No problem at all, just wanted to check in. Thanks for the reply.

KubaFYI committed:
shell: uart: Add option to route shell through USB CDC ACM

Addition of a simple way of permitting routing of shell through
USB CDC ACM UART.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

carlescufi commented:
@nordic-krch @jfischer-phytec-iot please re-review

carlescufi commented:
@KubaFYI can you fix compliance and CI (buildkite)?

carlescufi commented:
Relates #26818

pabigot commented on subsys/shell/shell_uart.c:
```diff
@@ -282,7 +288,9 @@ static int enable_shell_uart(struct device *arg)
 	return 0;
 }
-SYS_INIT(enable_shell_uart, POST_KERNEL, 0);
+SYS_INIT(enable_shell_uart, POST_KERNEL,
```
This might not be right.  If `CONFIG_UART_SHELL_ON_DEV_NAME` and `CONFIG_UART_CONSOLE_ON_DEV_NAME` are both true then the shell would be enabled on the device before the console is enabled (in level `APPLICATION` at priority `CONFIG_UART_CONSOLE_INIT_PRIORITY`).

pabigot commented on subsys/shell/shell_uart.c:
```diff
@@ -273,6 +275,10 @@ static int enable_shell_uart(struct device *arg)
 		CONFIG_LOG_MAX_LEVEL : CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL;
+	
+	if (IS_ENABLED(CONFIG_UART_SHELL_USE_CDC_ACM)) {
+		usb_enable(NULL);
```
As described for #27071 this hijacks a call that's supposed to be available to the application to register for USB event status changes.

pabigot reviewed (changes requested):
As-is this isn't going to work.  It needs to coordinate initialization with the console that may be on the same device; and it needs some way to function while still allowing the application to register for USB event state changes (which means it can't be installed until the USB subsystem has been enabled).

emob-nordic commented:
Closing as we agreed not to allow to call `usb_enable` form outside of the application. This will break the requirement mentioned in [2.2 release notes](https://github.com/zephyrproject-rtos/zephyr/blame/master/doc/releases/release-notes-2.2.rst#L158-L165) and [2.4 release notes](https://github.com/zephyrproject-rtos/zephyr/blame/master/doc/releases/release-notes-2.4.rst#L325-L336)

mniestroj commented:
@emob-nordic Then what is the recommended way to having shell over USB in Zephyr application?

pabigot commented:
I use a call to `wait_for_console()` in `main()`.  Something like this might be added to Zephyr as a helper function.

```
int
wait_for_console()
{
  struct device *cons = device_get_binding(console_label);
  int rc = 0;

  if (IS_ENABLED(CONFIG_USB)) {
    usb_enable(usb_dc_status);
  }

  if ((rc == 0) && IS_ENABLED(CONFIG_USB_UART_CONSOLE)) {
    uint32_t dtr;
    do {
      uart_line_ctrl_get(cons, UART_LINE_CTRL_DTR, &dtr);
    } while (!dtr);
  }

  printk("%s Up and running: %d\n", zxxpab::utils::now_str(), rc);

  return rc;
}
```
